src/program_iterator.jl

```diff
@@ -67,56 +67,112 @@ end
-            Expr(:kw, :(max_depth::Int), typemax(Int)), 
-            Expr(:kw, :(max_size::Int), typemax(Int)), 
-            Expr(:kw, :(max_time::Int), typemax(Int)), 
-            Expr(:kw, :(max_enumerations::Int), typemax(Int)),
-            Expr(:kw, :(solver::Union{Solver, Nothing}), nothing)
-        ]
+        kwargs_fields = map(esc, filter(is_kwdef, extrafields))
+        notkwargs     = map(esc, filter(!is_kwdef, extrafields))
+
+        # create field names
+        field_names = map(extract_name_from_argument, extrafields)
+
+        # throw an error if user used one of the reserved arg names 
+        RESERVERD_ARG_NAMES = [:solver,:start_symbol,:initial_node,:grammar,:max_depth,:max_size]
+        for field_name ∈ field_names
+            if field_name ∈ RESERVERD_ARG_NAMES
+                throw(ArgumentError(
+                    "When using the @programiterator macro you are not allowed to use any of the $RESERVERD_ARG_NAMES field names.
+                     This is because there would be conflicting names in the function signature. 
+                     However, '$field_name' was found as an argument name. 
+                     Please change the name of the field argument to not collide with the reserved argument names above.
+                    "))
+            end
+        end
 
+        field_names = map(esc, field_names)
+        escaped_name = esc(name) # this is the name of the struct
+
+        # keyword arguments come after the normal arguments (notkwargs)
+        all_constructors = Base.remove_linenums!(
+            :(
+              begin 
+                # solver with grammar and start symbol
+                function $(escaped_name)(grammar::AbstractGrammar, start_symbol::Symbol, $(notkwargs...) ; 
+                                        max_size = typemax(Int), max_depth = typemax(Int), $(kwargs_fields...) )
+                    return $(escaped_name)(GenericSolver(grammar, start_symbol, max_size = max_size, max_depth = max_depth), $(field_names...))
+                end
+
+                # solver with grammar and initial rulenode to start with
+                function $(escaped_name)(grammar::AbstractGrammar, initial_node::RuleNode, $(notkwargs...) ;
+                                        max_size = typemax(Int), max_depth = typemax(Int), $(kwargs_fields...) )
+                    return $(escaped_name)(GenericSolver(grammar, initial_node, max_size = max_size, max_depth = max_depth), $(field_names...))
+                end
+              end
+            )
+        )
+        # this constructor should ONLY be used when there are kwarg fields 
+        # otherwise this will overwrite the default julia struct constructor
+        solver_constructor = Base.remove_linenums!(:(
+          # solver main constructor
+          function $(escaped_name)(solver::Solver, $(notkwargs...) ; $(kwargs_fields...) )
+              return $(escaped_name)(solver, $(field_names...))
+          end
+        ))
```

Comment:
Does the following constructor also work?
```julia
solver = GenericSolver(grammar, :S, max_size=10, max_depth=5)
iterator = BFSIterator(solver)
```

**About the kwargs issue**: We could include `max_size` and `max_depth` as kwargs for the `solver_constructor` too. By default, the max_size and max_depth of the solver will be used. But if the user passes a specific max_size or max_depth to the iterator, they will be overwritten. For example:

```julia
solver = GenericSolver(grammar, :S, max_size=10, max_depth=5)
iterator = BFSIterator(solver, max_depth=3) #will overwrite solver.max_depth from 5 to 3. But keeps solver.max_size=10.
```

Reply:
The first one works nicely without issues as seen by this test
```jl
    @testset "Check default constructors with a solver" begin 
        solver = GenericSolver(g, :R, max_size=10, max_depth=5)
        iterator = BFSIterator(solver)
        @test get_grammar(iterator.solver) == g 
        @test get_max_size(iterator.solver) == 10 
        @test get_max_depth(iterator.solver) == 5
    end
 ```
 
 For the other one I had to make `solver` a kwarg in order to make julia figure out how to differentiate between the default constructor `Iterator(solver)` and `Iterator(solver; max_depth, max_size)`
 
```jl
    @testset "Check if max_depth and max_size are overwritten" begin 

        solver = GenericSolver(g, :R, max_size=10, max_depth=5)
        @test solver.max_size == 10
        @test solver.max_depth == 5
        # will overwrite solver.max_depth from 5 to 3. But keeps solver.max_size=10.
        iterator = BFSIterator(solver = solver, max_depth=3) # <- solver as kwarg
        @test get_max_size(solver) == 10 
        @test get_max_depth(solver) == 3
    end
    ```